snap/info_snap_yaml.go

```diff
@@ -618,3 +644,57 @@ func convertToSlotOrPlugData(plugOrSlot, name string, data interface{}) (iface,
 		return "", "", nil, err
 	}
 }
+
+// Short form:
+//   system-usernames:
+//     snap_daemon: shared  # 'scope' is 'shared'
+//     lxd: external        # currently unsupported
+//     foo: private         # currently unsupported
+// Attributes form:
+//   system-usernames:
+//     snap_daemon:
+//       scope: shared
+//       attrib1: ...
+//       attrib2: ...
+func convertToUsernamesData(user string, data interface{}) (scope string, attrs map[string]interface{}, err error) {
+	switch data.(type) {
+	case string:
+		return data.(string), nil, nil
+	case nil:
+		return "", nil, nil
+	case map[interface{}]interface{}:
+		for keyData, valueData := range data.(map[interface{}]interface{}) {
+			key, ok := keyData.(string)
+			if !ok {
+				err := fmt.Errorf("%q has attribute that is not a string (found %T)", user, keyData)
+				return "", nil, err
+			}
+			switch key {
+			case "scope":
+				value, ok := valueData.(string)
+				if !ok {
+					err := fmt.Errorf("scope on %q is not a string (found %T)", user, valueData)
+					return "", nil, err
+				}
+				scope = value
+			default:
+				// for maximum future compatibility, we allow
+				// value to be empty, but not key
+				if key != "" {
```

Comment:
not quite sure why we shouldn't make key == "" an error instead of just ignoring

Reply:
@pedronis - happy to make this an error, but this was what I understood to be the implementation based on feedback from recent conversations where we should not fail hard if a snap uses something else (sorta like when a snap specifies an interface the snapd on the system doesn't yet support).

Reply:
@pedronis - from our yaml doc: "IMPORTANT: Please teach snapd to recognize that as an unsupported extension from day one, instead of exploding (so we can do backwards compatible logic later). It should be parseable, but fail later that it is not supported yet"

Reply:
I'm not sure how a empty key in a map is a useful thing to support at any point? I'm missing something

Reply:
Ok, I see what you are getting at here. I'll adjust.